crates/language/src/language_settings.rs

```diff
@@ -802,6 +802,24 @@ pub struct InlayHintSettings {
     pub scroll_debounce_ms: u64,
 }
 
+#[derive(Clone, Default, Serialize, Deserialize, JsonSchema)]
+pub struct InlayHintSettingsContent {
+    /// Whether or not to enable inlay hints in the editor.
+    ///
+    /// Default: false
+    pub enabled: Option<bool>,
```

Comment:
We already have https://github.com/zed-industries/zed/blob/81ca890a2a65881b4526faeb4526b9c14c1228d5/crates/language/src/language_settings.rs#L760-L765 so it seems wrong to add another global thing with the same key and json key.

Reply:
I modeled this the same way `TabBarSettings` are handled here:

https://github.com/zed-industries/zed/blob/75c545aa1e7a9cb01febf2f6dc00536c7271ff2f/crates/workspace/src/workspace_settings.rs#L133-L149

Code for updating `TabBarSettings` was in the same file (`editor.rs`) and so it was easy to mimic already existing pattern.

I also thought about using `InlayHintSettings` as is but decided not to because it exposes much more fields than simply `enable`. There is no way to control them from the UI and so I feel wrong to use them as well.

If this approach is not the way to go for new code, I can redo it like you suggest, no problem.

Reply:
I am not sure what the right solution to all this, but there's one invariant we can hardly violate: `settings.json` already have `inlay_hints` key and a corresponding Rust settings struct associated with it, that is able to enable and disable hints globally and per language.

Also, unlike the tab bar which is sort of "global", inlays are language-specific, so that json may be tricky, something like

```json5
{
  "inlay_hints": {
    "enabled": false // disabled globally
  },
  "languages": {
    "Rust": {
      "inlay_hints": {
        "enabled": true // enabled for Rust only
      }
    }
  }
}
```

So, before trying to copy anything and write some code, I'd stop and understand what are we supposed to do with all this.

Adding a new setting group for inlays with the same key will never work, just due to the fact it conflicts with the pre-existing one.
Also that new value, even hidden under a different key, has to somehow interact with both the global and local language settings — do we really want to have another boolean in that picture?

Reply:
I've came up with this: 

```rust
let Some(workspace) = self.workspace() else {
    return;
};
let fs = workspace.read(cx).app_state().fs.clone();
let inlay_hints_enabled = self.inlay_hint_cache.enabled;
update_settings_file::<AllLanguageSettings>(fs, cx, move |settings, _| {
    let mut inlay_hints_settings =
        settings.defaults.inlay_hints.unwrap_or(InlayHintSettings {
            enabled: false,
            show_type_hints: true,
            show_parameter_hints: true,
            show_other_hints: true,
            show_background: false,
            edit_debounce_ms: 700,
            scroll_debounce_ms: 50,
        });
    inlay_hints_settings.enabled = inlay_hints_enabled;
    settings.defaults.inlay_hints = Some(inlay_hints_settings);
});
```

This attempt doesn't introduce another struct nor does it re-use `inlay_hints` key, but has 2 issues: 
1. It requires creating default value manually. Is there a better way to do that? Maybe make a copy from default settings?
2. All variables of `InlayHintSettings` struct are required and thus they will be serialized into user `settings.json`  despite not being configurable by user in the UI.

I think we should ignore languages-specific settings in this PR because we're unable to configure them in the UI.

Reply:
i think there's even worse issue besides all that: the config from my example in https://github.com/zed-industries/zed/pull/20974#discussion_r1851993306 actually enables hints for Rust.

So, changing the defaults won't work as it's not enough given that there could be a language override.
Also, as you have mentioned correctly, fiddling with json and adding arbitrary values is a problem.

---------------------

I strongly insist on

> I'd stop and understand what are we supposed to do with all this.

Without even touching the code, can you describe the behavior you want to achieve?
Here are my loose thoughts (which I really expected not to make, as hoped for a good PR that considered that, so did not think of it at all hence could not write them before. They also may be wrong, as I did not think much on them):

Invariant 1. There's a button in the UI that allows to turn hints on and off + there's a set of configs, the default one + the user one, each of them potentially containing a global hints override and a number of language-specific ones.

Invariant 2: User settings override the defaults; language-local settings override the globals.

Invariant(?) 3: Based on our discussion above, it's very complex to adjust json files programmatically. Maybe it's not even possible to do this correctly.

Feature description: When the user presses the toggle hints button, it has to reverse the final settings and show/hide the hints for the specific button.

Questions on that description: 
* does this have to be permanent per Zed session or between restarts?
(IMO, it should be the former, while you push for the latter, why, given that there are JSON settings for that already?)

* Toggling happens in a particular editor, which is language-specific (as we show the language name in the footer).
Should the toggle be global or per language? What if buffer's language changes?

* if it should be permanent between restarts, how is it different from adjusting the settings manually? We rarely do something like this (inline assistant is one case, is there anything else?) How are we supposed to update the JSON correctly, given all written above (e.g. in order to disable the settings globally you'll need to go over each language config and set it to false)

* if it should be non-permanent, after all, seems that there's a bug in the inlay hint cache implementation that resets the setting, rather than keeping it? If we should disable per language only, the whole inlay cache disabling is utterly wrong, as there's no language there at all? Also, the cache is _per editor_ so it's never going to work with the current design.

* something else you can add on the topic?

-------------------

By considering all this, I think that we should not adjust JSON settings at all, keeping the feature toggle in memory + store it elsewhere outside of the inlay hint cache.

So, there could be some global state, which we need to update.
We have https://github.com/zed-industries/zed/blob/9c3f64271dc4159572dcb8e29ed1fd97d3f93c68/crates/gpui/src/global.rs#L48-L56 and seems that we need to have some `HashMap<LanguageName, Foo>` for that?

This has to be accessed in the inlay_cache before doing anything and observer for changes: https://github.com/zed-industries/zed/blob/9c3f64271dc4159572dcb8e29ed1fd97d3f93c68/crates/gpui/src/window.rs#L3614

And current inlay_cache design with reacting to settings is wrong and has to be removed.

I might be wrong, but adjusting JSON seems like a total disaster to me, hence not a good alternative anyway.

Reply:
> Without even touching the code, can you describe the behavior you want to achieve?

I believe settings that are exposed in the UI should be persistent across between restarts. This is how software is generally built, and it would be a challenge to find something that doesn't conform to this.

I don't enjoy fiddling with configs manually, and so I went to see if there was an issue for this already raised. There was an issue (linked in the OP) and Joseph in his reply (https://github.com/zed-industries/zed/issues/15909#issuecomment-2278726499) mentions that this setting probably should be persistent. I don't know what role they have in the project or how much they have weight on any of this, but lack of responses from Zed staff sort of confirmed that this the way to go.

And so I went to see how it can be done. This PR is also motivated by settings UI goal in the [roadmap](https://zed.dev/roadmap). I thought I could do something useful for me, for person who raised the issue and for Zed at large. If these changes were to go through, I'd also plumb remaining settings in the same menu bringing Zed a little bit closer to this goal.

> does this have to be permanent per Zed session or between restarts?

I believe this should be a permanent one.

> Toggling happens in a particular editor, which is language-specific (as we show the language name in the footer).

This is how you (person working on the project full time) think about it. The way I think about it and the way it is communicated in the UI is different:
* The button for the menu is detached from any UI hints about programming language. The language name is specified in the footer, but this button is in the header. You can't expect users to associate two different things in two different places with same concept.
* The tooltip for the button says "Editor Controls". To me an "editor" is the whole Zed, not a specific instance of file editing widget.
* I read it as a temporary stopgap until there is full-blown settings widget.

For the rest of your questions my opinion can be derived from the ones I have answered.

Personally I don't think it would be difficult to differentiate between "user globals" and "language specific" settings even at the current stage (both of UI and plumbing). For example, the button in the header would correspond to user globals and the on in the footer would be language specific. Yes, this would look rather stupid, but it is solvable now. To help  differentiate between the two it would make sense to use headers/sections like this one:
![image](https://github.com/user-attachments/assets/eb642bbf-a8cb-4d44-bbe8-d349ee842a3b)

Reply:
Ok, back to editing JSONs then: 

> How are we supposed to update the JSON correctly, given all written above (e.g. in order to disable the settings globally you'll need to go over each language config and set it to false)

Reply:
Based on snippet I sent today we would do the following:

1. Access `AllLanguageSettings` (or rather `AllLanguageSettingsContent`).
2. For user global set `default.inlay_hints.enabled` to `false`.
3. For language specific iterate over all `languages` and set `inlay_hints.enabled` to `false` for each of them.

`languages` collection holds only those languages for which there is configuration present. Furthermore we're working here with the `AllLanguageSettingsContent` struct that has all fields marked as optional. All this means is that we will only update configurations for languages which are explicitly present and have `inlay_hint` set.

Reply:
The snippet sent today had this under it:

> This attempt doesn't introduce another struct nor does it re-use `inlay_hints` key, but has 2 issues: 
> 1. It requires creating default value manually. Is there a better way to do that? Maybe make a copy from default settings?
> 2. All variables of `InlayHintSettings` struct are required and thus they will be serialized into user `settings.json`  despite not being configurable by user in the UI.
> 
> I think we should ignore languages-specific settings in this PR because we're unable to configure them in the UI.

did I miss anything, or this has to be solved either?

Reply:
Nope, you didn't miss anything. These problems are yet to be solved.

I think the plan would be:
1. Introduce `InlayHintSettingsContent` struct to hold same values as `InlayHintSettings` but as `Option`. This is duplication in its fines form, but the pattern is there and I don't know what else could be done. I don't know if Rust macros are advanced (or primitive enough) to generate two struct at the same time and keep them in sync, but in C++ it would be possible albeit ugly.
2. Replace `InlayHintSettings` with `InlayHintSettingsContent` in an `AllLanguageSettingsContent`.
3. ATM we can only change on/off state and so first initialization would be super simple:  `InlayHintSettingsContent` holds only optional values and we need to set only one.

Reply:
I'm sort of sad by the fact I have to write a 4th message in a row, torturing out of you all these details after saying

> I'd stop and understand what are we supposed to do with all this.

twice by that point.
I would really prefer to hear all that combined in one answer, even if that answer have to come N days later.
That makes me suspicious that more issues are to come, so I would expect these changes to come with tests, covering more than one scenario, as a compromise to all this.

One concern I do not know much about are plugins and the fact that they may come with more settings (those are not covering inlay hints now IIRC but later they might). 


Also, I am still salty about JSON-based persistence approach, given that it's based solemnly on https://github.com/zed-industries/zed/issues/15909#issuecomment-2278726499 comment from a non-developer (a brilliant non-developer, to say, but still...) that says `maybe` in that message + your own preference solemnly:

> I don't enjoy fiddling with configs manually

> I read it as a temporary stopgap until there is full-blown settings widget.

and the complexity this all brings.

But great, we have a plan now.

Reply:
another scenario to test: if config file is **read-only**, i.e. zed doesn't have permission to modify the config.